### PR TITLE
[generic] Extract subtitles from video.js

### PR DESCRIPTION
**Pull request by Lesmiscore**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [x] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

This adds support for extracting subtitles from video.js embeds.

Example URL where it happens: https://www.nhk.or.jp/d-garage/catchup/r/325245.html (yt-dlp also gets sources from development code, but whatever)

### CODE REVIEW

File: yt_dlp/extractor/generic.py [hunk 1]
```diff
@@ -3778,11 +3778,12 @@ def _real_extract(self, url):
...
         # Video.js embed
         mobj = re.search(
-            r'(?s)\bvideojs\s*\(.+?\.src\s*\(\s*((?:\[.+?\]|{.+?}))\s*\)\s*;',
+            r'(?s)\bvideojs\s*\(.+?([a-zA-Z0-9_$]+?)\.src\s*\(\s*((?:\[.+?\]|{.+?}))\s*\)\s*;',
             webpage)
         if mobj is not None:
+            varname = mobj.group(1)
             sources = self._parse_json(
-                mobj.group(1), video_id, transform_source=js_to_json,
+                mobj.group(2), video_id, transform_source=js_to_json,
```
Comment (pukkandan) [inline yt_dlp/extractor/generic.py:3786]: why?

File: yt_dlp/extractor/generic.py [hunk 2]
```diff
@@ -3819,6 +3820,24 @@ def _real_extract(self, url):
                             'Referer': full_response.geturl(),
                         },
                     })
+            # https://docs.videojs.com/player#addRemoteTextTrack
+            # https://html.spec.whatwg.org/multipage/media.html#htmltrackelement
+            for sub_match in re.finditer(rf'(?s){re.escape(varname)}' r'\.addRemoteTextTrack\(({.+?})\s*(?:,\s*(?:true|false))\)', webpage):
```
Comment (pukkandan) [inline yt_dlp/extractor/generic.py:3825]:
```suggestion
            for sub_match in re.finditer(rf'(?s){re.escape(varname)}' r'\.addRemoteTextTrack\(({.+?})\s*,\s*(?:true|false)\)', webpage):
```

File: yt_dlp/extractor/generic.py [hunk 3]
```diff
@@ -3819,6 +3820,24 @@ def _real_extract(self, url):
...
+                sub = self._parse_json(
+                    sub_match.group(1), video_id, transform_source=js_to_json,
+                    fatal=False)
+                if not sub or not isinstance(sub, dict):
+                    continue
+                src = sub.get('src')
+                if not src or not isinstance(src, str):
+                    continue
+                subtitles.setdefault(sub.get('language') or sub.get('srclang') or 'und', []).append({
```
Comment (pukkandan) [inline yt_dlp/extractor/generic.py:3834]:
```suggestion
                subtitles.setdefault(dict_get(sub, ('language', 'srclang')) or 'und', []).append({
```

File: yt_dlp/extractor/generic.py [hunk 4]
```diff
@@ -3819,6 +3820,24 @@ def _real_extract(self, url):
...
+            # https://html.spec.whatwg.org/multipage/media.html#htmltrackelement
+            for sub_match in re.finditer(rf'(?s){re.escape(varname)}' r'\.addRemoteTextTrack\(({.+?})\s*(?:,\s*(?:true|false))\)', webpage):
+                sub = self._parse_json(
+                    sub_match.group(1), video_id, transform_source=js_to_json,
+                    fatal=False)
+                if not sub or not isinstance(sub, dict):
+                    continue
+                src = sub.get('src')
+                if not src or not isinstance(src, str):
+                    continue
```
Comment (pukkandan) [inline yt_dlp/extractor/generic.py:3833]:
```suggestion
                sub = self._parse_json(
                    sub_match.group(1), video_id, transform_source=js_to_json, fatal=False) or {}
                src = str_or_none(sub.get('src')
                if not src:
                    continue
```

**Review by pukkandan — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/generic.py (hunk 1 above)
Comment (Lesmiscore) [reply yt_dlp/extractor/generic.py:3786]: To extract subtitles belonging to the same object

Comment (pukkandan) [reply yt_dlp/extractor/generic.py:3786]: 🤦 ignore me

File: yt_dlp/extractor/generic.py [hunk 5]
```diff
@@ -3778,11 +3778,12 @@ def _real_extract(self, url):
 
         # Video.js embed
         mobj = re.search(
-            r'(?s)\bvideojs\s*\(.+?\.src\s*\(\s*((?:\[.+?\]|{.+?}))\s*\)\s*;',
+            r'(?s)\bvideojs\s*\(.+?([a-zA-Z0-9_$]+?)\.src\s*\(\s*((?:\[.+?\]|{.+?}))\s*\)\s*;',
```
Comment (pukkandan) [inline yt_dlp/extractor/generic.py:3781]:
```suggestion
            r'(?s)\bvideojs\s*\(.+?([a-zA-Z0-9_$]+)\.src\s*\(\s*((?:\[.+?\]|{.+?}))\s*\)\s*;',
```

**Pull request merged**